spec/template.dd

```diff
@@ -1082,6 +1082,23 @@ $(H2 $(LNAME2 function-templates, Function Templates))
         }
         ---
 
+    $(P Variadic Function Templates can have parameters with default values.
+        These parameters are always set to their default value in case of IFTI.
+    )
+
+        $(SPEC_RUNNABLE_EXAMPLE_RUN
+        ---
+        size_t fun(T...)(T t, string file = __FILE__)
+        {
+            import std.stdio;
+            writeln(file, " ", t);
+            return T.length;
+        }
+        assert(fun(1, "foo") == 2);  // not 1 (IFTI)
+        assert(fun!int(1, "foo") == 1);  // IFTI
+        ---
```

Comment:
```
collect2: fatal error: ld terminated with signal 11 [Segmentation fault], core dumped
```

So that's weird. It compiles fine on my machine though with the `test_dspect` target I spuriously get the core dump.

`gdb` says:

```
Thread 6 (Thread 0x7ffff4e8f700 (LWP 13237)):
#0  0x00007ffff7bca73c in __lll_lock_wait () from /usr/lib/libpthread.so.0
#1  0x00007ffff7bc3ac6 in pthread_mutex_lock () from /usr/lib/libpthread.so.0
#2  0x00005555556ad179 in _D4core4sync5mutex5Mutex__T12lock_nothrowTCQBpQBnQBlQBiZQBdMFNbNiNeZv ()
#3  0x00005555556ad09d in core.sync.mutex.Mutex.lock() ()
#4  0x0000555555664f8f in _D3std11parallelism__T15ParallelForeachTSQBn9algorithm9iteration__T12FilterResultS_DQDe4file10dirEntriesFAyaQdEQEfQBb8SpanModebZ1fMFNaNbNfSQFhQCd8DirEntryZbTSQGaQCw11DirIteratorZQEhZQGe7opApplyMFMDFQCiZiZ4doItMFZ8makeTempMFZm (this=0x7ffff4e8e9f8)
    at /home/seb/dlang/phobos/std/parallelism.d-mixin-4011:4110
#5  0x0000555555664d64 in _D3std11parallelism__T15ParallelForeachTSQBn9algorithm9iteration__T12FilterResultS_DQDe4file10dirEntriesFAyaQdEQEfQBb8SpanModebZ1fMFNaNbNfSQFhQCd8DirEntryZbTSQGaQCw11DirIteratorZQEhZQGe7opApplyMFMDFQCiZiZ4doItMFZv (this=0x7fffffffd290)
    at /home/seb/dlang/phobos/std/parallelism.d-mixin-4011:4153
#6  0x00005555556b995c in _D3std11parallelism__T3runTDFZvZQkFQiZv ()
#7  0x00005555556b9498 in _D3std11parallelism__T4TaskSQBaQz3runTDFZvZQv4implFPvZv ()
#8  0x00005555556eef67 in std.parallelism.AbstractTask.job() ()
#9  0x00005555556b808c in _D3std11parallelism8TaskPool5doJobMFPSQBkQBj12AbstractTaskZv ()
#10 0x00005555556b81ea in std.parallelism.TaskPool.executeWorkLoop() ()
#11 0x00005555556b8191 in std.parallelism.TaskPool.startWorkLoop() ()
#12 0x00005555556d2d94 in core.thread.Thread.run() ()
#13 0x0000555555706788 in thread_entryPoint ()
#14 0x00007ffff7bc108c in start_thread () from /usr/lib/libpthread.so.0
#15 0x00007ffff6f89e7f in clone () from /usr/lib/libc.so.6

Thread 5 (Thread 0x7ffff5690700 (LWP 13236)):
#0  0x00007ffff7bcb260 in nanosleep () from /usr/lib/libpthread.so.0
#1  0x00005555556d2b52 in _D4core6thread6Thread5sleepFNbNiSQBf4time8DurationZv ()
#2  0x0000555555708228 in core.internal.spinlock.SpinLock.yield(ulong) shared ()
#3  0x00005555557081c4 in core.internal.spinlock.SpinLock.lock() shared ()
#4  0x00005555556dc781 in _D2gc4impl12conservativeQw14ConservativeGC__T9runLockedS_DQCeQCeQCcQCnQBs12mallocNoSyncMFNbmkKmxC8TypeInfoZPvS_DQEgQEgQEeQEp10mallocTimelS_DQFiQFiQFgQFr10numMallocslTmTkTmTxQCzZQFcMFNbKmKkKmKxQDsZQDl ()
#5  0x00005555556d60b2 in _D2gc4impl12conservativeQw14ConservativeGC6mallocMFNbmkxC8TypeInfoZPv ()
#6  0x00005555556ad61f in gc_malloc ()
#7  0x00005555556af3f9 in _d_allocmemory ()
#8  0x000055555565aabc in dspec_tester.main(immutable(char)[][]).__foreachbody3(std.file.DirEntry) (
    this=0x7fffffffd450, file=...) at dspec_tester.d:125
#9  0x0000555555664e13 in _D3std11parallelism__T15ParallelForeachTSQBn9algorithm9iteration__T12FilterResultS_DQDe4file10dirEntriesFAyaQdEQEfQBb8SpanModebZ1fMFNaNbNfSQFhQCd8DirEntryZbTSQGaQCw11DirIteratorZQEhZQGe7opApplyMFMDFQCiZiZ4doItMFZv (this=0x7fffffffd290)
    at /home/seb/dlang/phobos/std/parallelism.d-mixin-4011:4183
#10 0x00005555556b995c in _D3std11parallelism__T3runTDFZvZQkFQiZv ()
#11 0x00005555556b9498 in _D3std11parallelism__T4TaskSQBaQz3runTDFZvZQv4implFPvZv ()
#12 0x00005555556eef67 in std.parallelism.AbstractTask.job() ()
#13 0x00005555556b808c in _D3std11parallelism8TaskPool5doJobMFPSQBkQBj12AbstractTaskZv ()
#14 0x00005555556b81ea in std.parallelism.TaskPool.executeWorkLoop() ()
#15 0x00005555556b8191 in std.parallelism.TaskPool.startWorkLoop() ()

Thread 1 (Thread 0x7ffff7f9ca80 (LWP 13228)):
#0  0x00007ffff7bc9cf6 in do_futex_wait.constprop () from /usr/lib/libpthread.so.0
#1  0x00007ffff7bc9de8 in __new_sem_wait_slow.constprop.0 () from /usr/lib/libpthread.so.0
#2  0x0000555555706f26 in thread_suspendAll ()
#3  0x00005555556da3db in _D2gc4impl12conservativeQw3Gcx11fullcollectMFNbbZm ()
#4  0x00005555556d8b7e in _D2gc4impl12conservativeQw3Gcx8bigAllocMFNbmKmkxC8TypeInfoZPv ()
#5  0x00005555556dc7d9 in _D2gc4impl12conservativeQw14ConservativeGC__T9runLockedS_DQCeQCeQCcQCnQBs12mallocNoSyncMFNbmkKmxC8TypeInfoZPvS_DQEgQEgQEeQEp10mallocTimelS_DQFiQFiQFgQFr10numMallocslTmTkTmTxQCzZQFcMFNbKmKkKmKxQDsZQDl ()
#6  0x00005555556d60b2 in _D2gc4impl12conservativeQw14ConservativeGC6mallocMFNbmkxC8TypeInfoZPv ()
#7  0x00005555556ad61f in gc_malloc ()
#8  0x00005555556aca94 in core.memory.GC.malloc(ulong, uint, const(TypeInfo)) ()
#9  0x00005555556e472f in _D3std5array__T14arrayAllocImplVbi0TAhTmZQBaFNaNbmZQp ()
#10 0x00005555556e46d1 in _D3std5array__T18uninitializedArrayTAhTymZQBbFNaNbNeymZQt ()
#11 0x00005555556b53d3 in std.file.readImpl(const(char)[], const(char)*, ulong) ()
#12 0x0000555555665c9a in _D3std4file__T4readTAyaZQkFNfQjmZAv (upTo=18446744073709551615, name=...)
    at /home/seb/dlang/phobos/std/file.d:300
#13 0x0000555555665bb8 in _D3std4file__T8readTextTAyaTQeZQrFNfQmZQp (name=...)
    at /home/seb/dlang/phobos/std/file.d:491
#14 0x0000555555665b90 in _D3std4file__T8readTextTAyaTSQBbQBa8DirEntryZQBfFNfKQyZQBf (name=...)
    at /home/seb/dlang/phobos/std/file.d:510
#15 0x0000555555665b14 in _D12dspec_tester4mainFAAyaZ__T9__lambda2TS3std4file8DirEntryTQBmZQBjFNfQBeQBzZSQBl9algorithm9iteration__T9MapResultSQEk15ddocMacroToCodeTSQDsQChQCa__TQBtSQFw23untilClosingParenthesesTSQFmQEbQDu__T8splitterVQHba6_61203d3d2062TQHuTQHyZQBlFQIgQIjZ6ResultZQFrZQFv (__HID10=0x7fffffffbad0, 
    ddocKey=..., file=...) at dspec_tester.d:115
``

I'm not entirely sure what's going on. Is the GC trying to collect the `uninitializedArray` (?), but it's definitely an issue with the spec tester and not your fault. Sorry for asking you to try an experimental feature.

Reply:
side question: why is the backtrace only partially demangled even though:
```
 echo _D3std5array__T14arrayAllocImplVbi0TAhTmZQBaFNaNbmZQp|ddemangle
pure nothrow ubyte[] std.array.arrayAllocImpl!(false, ubyte[], ulong).arrayAllocImpl(ulong)
```

Reply:
filed https://issues.dlang.org/show_bug.cgi?id=18490 so we don't forget